### ISSUE TITLE
Navigate to nearest point in track

### ISSUE THREAD

**Issue by algil** [Falcosc/locus-addon-tasker#27]
Hi,
Currently the Locus API and Locus/Tasker plugin provide Action Tasks "Navigation/Recalculate" (current track) and "Navigate To" (a new target point). However more useful for me would be "Navigate/Nearest point" (current track).  When navigating a track if I decide to shortcut/bypass part of the track then Locus continues to target the section of the track bypassed - I think it will not jump ahead more than 100 track points, so need then to use the "Navigate/Nearest point" action.  

Is there any way to get this added ?  If I raise an Idea in Locus myself I'm doubtful there would ever be enough API users to get enough votes.
Thanks
Alan

**Comment by Falcosc** [comment]
It is very expensive to find the nearest point of a large point list if the nearest point could be located in every direction around your local position.

You would need to limit this search operation to a small number of points.

I guess you don't want to use the "next Navigation Point"? The only smart way of reducing the time to wait for finding the next point would be a configurable limit of points after and before the next navigation point.

For geotagging it was a bit less difficult because finding the nearest timestamp was a one dimensional task and binary search was great for that: https://github.com/Falcosc/locus-addon-tasker/blob/1433b019b256a911283e22ea6be752e3877e4625/app/src/main/java/falcosc/locus/addon/tasker/GeotagPhotosService.java#L312
With 2 dimensions, I don't know, you can't even sort the points because of having 2 dimensions. This would be a great question for someone, how did study such kind problems.

**Comment by algil** [comment]
Hi 
I am not suggesting to calculate the nearest point in the Locus/tasker plugin.   I only want to be able to instruct Locus to execute the "Nearest point" action in a similar way as is available for "Recalculate".  "Nearest Point" action causes Locus to set the target point on the track as the nearest track point to the current location.

So, when navigating a track in Locus tapping on the navigation arrow allows to then choose "Nearest point" or "Recalculate".   The Locus/tasker plugin allows to automate "Recalculate" by executing Action Task "Navigation/Recalculate".   I would like there to be also an Action Task in the Locus/tasker plugin for "Navigation/Nearest point"  to automate this also.  
As far as I can see, the Locus API doesn't support this so it would need a change in the Locus API as well as the plugin.  

Currently I'm forced as a workaround to use the Tasker/AutoInput plugin to press the buttons in Locus but its not ideal.

**Comment by Falcosc** [comment]
Jumping forward is more easy, but you are right, there is no way to tell locus to execute the existing action. But maybe it is easy to integrate in the action tasks because it is an existing function.

**Comment by algil** [comment]
Yes, I'm hoping the Locus team could add it without much effort. Not sure how to take that forward - if I raise as an Idea it probably wouldn't get many votes.  Or should I raise it on the forum first ?

**Comment by algil** [comment]
Locus Map has been updated to include broadcast Action Task 
navigation
action: "nearest_point"

See:  https://help.locusmap.eu/topic/26875-broadcast-intent-navigation-to-nearest-point

It's currently in beta  LM 4.28.0.1_1164,  Not sure if any changes to the API itself involved  .  Anyway, v4.28  should be released fairly soon I think, 
Though you can invoke the Intent directly in Tasker, it would be nice to also include this in the plugin.

**Comment by Falcosc** [comment]
Would you like to compare https://github.com/asamm/locus-api/wiki/Action-tasks-(Broadcasts) with the current action tasks?

I didn't write down which revision was implemented.
But I can tell that all revision after the last release in 2011 will be missing https://github.com/asamm/locus-api/wiki/Action-tasks-(Broadcasts)/_compare/ea141a8048770a64e6f3b22d93ac4c13065eb5af...aa13f96ff0ddcc475b393632f0f4707db816b46d

But maybe we miss changes from 2011 as well and thats why I am asking for a check.

**Comment by Falcosc** [comment]
I should have found all differences.
Would you like to test the changes?

New Tasks gps_on_off, map_layer_base, map_overlay, map_reload_theme, weather. And parameters: add_wpt.description, navigation.nearest_point

use this link to join or leave the beta program
https://play.google.com/apps/testing/falcosc.locus.addon.tasker

**Comment by algil** [comment]
I compared also and I get same as you, plus possibly 
map_move_x,   map_move_y,  and   map_move_zoom     new option "keep centering"  ?
screen_lock new option "toggle"

I also saw some time ago in the Forum another Intent for adding a point to the map. It doesn't appear to be in the API documentation.
It was added in V4.24.0, changelog entry:  "add: sending of points over broadcasts (experimental)". For details see:-
https://forum.locusmap.eu/index.php?topic=8922.msg76051#msg76051
I tested this now and it works. 

Yes I will test any changes you make.

**Comment by Falcosc** [comment]
Adding data is already exposed via the API methods, in the end these methods just generate intends. So yes, there are a lot of intends which are not related to action tasks.

Let's keep "keep centering" exclusive for "Map Move Zoom"
Thank you, didn't catch that

**Comment by algil** [comment]
I tested build 78 using Locus Map beta v4.28.0.1.

These were OK:  gps_on_off, map_layer_base, map_overlay,  track_record/add_wpt.description,   navigation.nearest_point,  map_move_zoom "keep centering" , screen_lock.toggle

I could not test "map_reload_theme" as I have no knowledge of editing themes.  I think Tapio is a theme creator so he might be able to test. 
---------

"weather" was not OK - nothing happens in Locus, though I believe this is because of an issue in Locus itself, not the plugin -  I also tried it as direct Send Intent from tasker and got the same. If you could also test it and if you get the same then I will raise a problem ticket with the LM support desk

**Comment by algil** [comment]
The Action Task "weather" in fact does work but you have to add   action: "start". The documentation is not quite correct.
So the Extra would be:
tasks:{weather:{action:"start"}}    - shows the weather at the map center coordinates.
tasks:{weather:{action:"start", lat:"NN.NNNNN", lon:"NN.NNNNN"}}   - show weather at provided coordinates.

It does not appear to support "stop" or "toggle".
